signer/src/roinput.rs

```diff
@@ -0,0 +1,867 @@
+/// its `to_roinput()` serialization method using the [ROInput] functions below.
+///
+/// For example,
+///
+/// ```rust
+/// use mina_signer::{CompressedPubKey, Hashable, NetworkId, ROInput, Signable};
+///
+/// #[derive(Clone, Copy)]
+/// pub struct MyExample {
+///     pub account: CompressedPubKey,
+///     pub amount: u64,
+///     pub nonce: u32,
+/// }
+///
+/// impl Hashable for MyExample {
+///     fn to_roinput(self) -> ROInput {
+///         let mut roi = ROInput::new();
+///
+///         roi.append_field(self.account.x);
+///         roi.append_bit(self.account.is_odd);
+///         roi.append_u64(self.amount);
+///         roi.append_u32(self.nonce);
+///
+///         roi
+///     }
+/// }
+/// ```
+/// **Details:** For technical reasons related to our proof system and performance, fields are
+/// serialized for signing differently than other types. Additionally, during signing all members
+/// of the random oracle input get serialized together in two different ways: both as *bytes* and
+/// as a vector of *field elements*.  The random oracle input encapsulates and automates this
+/// complexity.
+
+#[derive(Default)]
+pub struct ROInput {
+    fields: Vec<BaseField>,
+    bits: BitVec<Lsb0, u8>,
+}
+
+impl ROInput {
+    /// Create a new empty random oracle input
+    pub fn new() -> Self {
+        ROInput {
+            fields: vec![],
+            bits: BitVec::new(),
+        }
+    }
+
+    /// Append a base field element
+    pub fn append_field(&mut self, f: BaseField) {
+        self.fields.push(f);
+    }
+
+    /// Append a scalar field element
+    pub fn append_scalar(&mut self, s: ScalarField) {
+        // mina scalars are 255 bytes
+        let bytes = s.to_bytes(); // TODO: Combine these two into one-liner
+        let bits = &bytes.as_bits::<Lsb0>()[..ScalarField::size_in_bits()];
+        self.bits.extend(bits);
+    }
+
+    /// Append a single bit
+    pub fn append_bit(&mut self, b: bool) {
+        self.bits.push(b);
+    }
+
+    /// Append bytes
+    pub fn append_bytes(&mut self, bytes: &[u8]) {
+        self.bits.extend_from_bitslice(bytes.as_bits::<Lsb0>());
+    }
+
+    /// Append a 32-bit unsigned integer
+    pub fn append_u32(&mut self, x: u32) {
+        self.append_bytes(&x.to_le_bytes());
+    }
+
+    /// Append a 64-bit unsigned integer
+    pub fn append_u64(&mut self, x: u64) {
+        self.append_bytes(&x.to_le_bytes());
+    }
+
+    /// Serialize random oracle input to bytes
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bits: BitVec<Lsb0, u8> = self.fields.iter().fold(BitVec::new(), |mut acc, fe| {
+            acc.extend_from_bitslice(&fe.to_bytes().as_bits::<Lsb0>()[..BaseField::size_in_bits()]);
+
+            acc
+        });
+
+        bits.extend(&self.bits);
+
+        bits.into()
+    }
+
+    /// Serialize random oracle input to vector of base field elements
+    pub fn to_fields(&self) -> Vec<BaseField> {
```

Comment:
I think I pointed that before, but this will have collisions if you don't include the length of bits somewhere

Reply:
Yes, but I can't change this or it will not be compatible with existing mainnet/testnet signatures.  If we want to fix this we should introduce tx version 2 and definitely pad with the length.